speculator/train_speculator_utils.py

```diff
@@ -0,0 +1,546 @@
+                speculator,
+                base_model_input,
+                input,
+                loss_fn,
+                ddp_stats,
+                base_model_mesh,
+            )
+
+        loss.backward()
+        ddp_stats[0] += speculator.clip_grad_norm_(cfg.grad_clip_thresh).item()
+        optimizer.step()
+        scheduler.step()
+
+        ddp_stats[1] += 1
+
+        if profiler:
+            profiler.step()
+
+        if batch_idx % cfg.report_interval == 0:
+            dist.all_reduce(ddp_stats, op=dist.ReduceOp.SUM)
+            train_loss = ddp_stats[2:] / ddp_stats[1]
+            g_norm = ddp_stats[0] / ddp_stats[1]
+            elapsed_time = time.time() - loop_start
+            world_size = int(os.environ["WORLD_SIZE"])
+            elapsed_tokens += cfg.report_interval * world_size * step_tok
+            if rank == 0:
+                print(f"{time.time()}")
+                print("step:", batch_idx)
+                print("tokens seen:", n_tok + elapsed_tokens)
+                for i in range(len(train_loss)):
+                    print(f"loss {i+1}:", train_loss[i].item())
+                print("gradient norm:", g_norm.item())
+                print(
+                    f"speed for these {cfg.report_interval} steps:",
+                    (time.time() - start) / cfg.report_interval,
+                )
+                print("overall speed:", elapsed_time / (batch_idx - start_step))
+                print("LR:", scheduler.get_last_lr())
+                print(
+                    "reserved memory:",
+                    torch.cuda.max_memory_reserved(device=torch.cuda.current_device()),
+                )
+                print(
+                    "active memory:",
+                    torch.cuda.max_memory_allocated(device=torch.cuda.current_device()),
+                )
+                print(
+                    "overall token per gpu per sec:",
+                    int(elapsed_tokens / world_size / elapsed_time),
+                )
+                print("token per day:", int(elapsed_tokens / elapsed_time * 3600 * 24))
+                print()
+            start = time.time()
+            ddp_stats.zero_()
+        torch.cuda.reset_peak_memory_stats(device=torch.cuda.current_device())
+
+        if (
+            batch_idx % cfg.checkpoint_interval == 0
+            or do_ckpt(cfg.ckpt_save_path) is True
+        ):
+            torch.cuda.empty_cache()
+            checkpointer.save(
+                batch_idx,
+                speculator,
+                optimizer,
+                train_loader,
+                tokens_seen=elapsed_tokens + n_tok,
+            )
+            torch.cuda.empty_cache()
+            do_ckpt(cfg.ckpt_save_path, reset=True)
+
+    checkpointer.save_single_file(
+        batch_idx,
+        speculator,
+        tokens_seen=elapsed_tokens + n_tok,
+        is_compiled=cfg.use_torch_compile,
+    )
+
+    return train_loss
+
+
+class EmbedGPTBigCode(GPTBigCode):
```

Comment:
may want to remove these in favor of a HiddenStatesExtractor wrapper, but not crucial for now. This way we can remove all the extra variants and adapters.